step/step.go

```diff
@@ -0,0 +1,553 @@
+
+	// Device Destination
+	deviceDestination := fmt.Sprintf("id=%s", sim.ID)
+
+	s.logger.Printf("* device_destination: %s", deviceDestination)
+	s.logger.Println()
+
+	if input.TestRepetitionMode != testRepetitionModeNone && xcodeMajorVersion < 13 {
+		return Config{}, errors.New("Test Repetition Mode (test_repetition_mode) is not available below Xcode 13")
+	}
+
+	if input.TestRepetitionMode != testRepetitionModeNone && input.MaximumTestRepetitions < 2 {
+		return Config{}, fmt.Errorf("invalid number of Maximum Test Repetitions (maximum_test_repetitions): %d, should be more than 1", input.MaximumTestRepetitions)
+	}
+
+	if input.RelaunchTestsForEachRepetition && input.TestRepetitionMode == testRepetitionModeNone {
+		return Config{}, errors.New("Relaunch Tests for Each Repetition (relaunch_tests_for_each_repetition) cannot be used if Test Repetition Mode (test_repetition_mode) is 'none'")
+	}
+
+	if input.RetryTestsOnFailure && xcodeMajorVersion > 12 {
+		return Config{}, errors.New("Should retry tests on failure? (should_retry_test_on_fail) is not available above Xcode 12; use test_repetition_mode=retry_on_failure instead")
+	}
+
+	return Config{
+		ProjectPath: projectPath,
+		Scheme:      input.Scheme,
+		TestPlan:    input.TestPlan,
+
+		XcodeMajorVersion: int(xcodeMajorVersion),
+		SimulatorID:       sim.ID,
+		IsSimulatorBooted: sim.Status != simulatorShutdownState,
+
+		TestRepetitionMode:            input.TestRepetitionMode,
+		MaximumTestRepetitions:        input.MaximumTestRepetitions,
+		RelaunchTestForEachRepetition: input.RelaunchTestsForEachRepetition,
+
+		OutputTool:         input.OutputTool,
+		IsCleanBuild:       input.IsCleanBuild,
+		IsSingleBuild:      input.IsSingleBuild,
+		BuildBeforeTesting: input.ShouldBuildBeforeTest,
+
+		RetryTestsOnFailure:       input.RetryTestsOnFailure,
+		DisableIndexWhileBuilding: input.DisableIndexWhileBuilding,
+		GenerateCodeCoverageFiles: input.GenerateCodeCoverageFiles,
+		HeadlessMode:              headlessMode,
+
+		XcodebuildTestOptions: input.TestOptions,
+		XcprettyOptions:       input.XcprettyTestOptions,
+
+		SimulatorDebug: simulatorDebug,
+
+		DeployDir:             input.DeployDir,
+		ExportUITestArtifacts: exportUITestArtifacts,
+
+		CacheLevel: input.CacheLevel,
+	}, nil
+}
+
+// InstallDeps ...
+func (s XcodeTestRunner) InstallDeps(xcpretty bool) error {
+	if !xcpretty {
+		return nil
+	}
+
+	xcprettyVersion, err := s.xcprettyInstaller.Install()
+	if err != nil {
+		return fmt.Errorf("an error occured during installing xcpretty: %s", err)
+	}
+	s.logger.Printf("- xcprettyVersion: %s", xcprettyVersion.String())
+	s.logger.Println()
+
+	return nil
+}
+
+// Result ...
+type Result struct {
+	XcresultPath             string
+	XcodebuildBuildLog       string
+	XcodebuildTestLog        string
+	SimulatorDiagnosticsPath string
+}
+
+// Run ...
+func (s XcodeTestRunner) Run(cfg Config) (Result, error) {
```

Comment:
I'd somehow split this into more methods like start simulator, execute build, execute test, etc.

Reply:
Broken into: `prepareSimulator`, `runTests` and `teardownSimulator` in https://github.com/bitrise-steplib/steps-xcode-test/pull/181/commits/78e176b4d19dfa9dad36ae8e232df40c5749fac4 .

`CollectSwiftPackages` stayed inline.